datavec/datavec-python/src/main/java/org/datavec/python/PythonType.java

```diff
@@ -0,0 +1,175 @@
+/*******************************************************************************
+ * Copyright (c) 2019 Konduit K.K.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License, Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ******************************************************************************/
+
+package org.datavec.python;
+
+import lombok.Data;
+import org.bytedeco.javacpp.BytePointer;
+import org.nd4j.linalg.api.ndarray.INDArray;
+
+import java.util.List;
+import java.util.Map;
+
+import static org.datavec.python.Python.importModule;
+
+
+public abstract class PythonType<T> {
+
+    public abstract T toJava(PythonObject pythonObject) throws PythonException;
+    public abstract TypeName getName();
+    public String toString(){
+        return getName().name();
+    }
+    enum TypeName{
+        STR,
+        INT,
+        FLOAT,
+        BOOL,
+        LIST,
+        DICT,
+        NDARRAY,
+        BYTES
+    }
+
+    public static PythonType valueOf(String typeName) throws PythonException{
+        try{
+            typeName.valueOf(typeName);
+        } catch (IllegalArgumentException iae){
+            throw new PythonException("Invalid python type: " + typeName, iae);
+        }
+        try{
+            return (PythonType)PythonType.class.getField(typeName).get(null); // shouldn't fail
+        } catch (Exception e){
+            throw new RuntimeException(e);
+        }
+
+    }
+    public static PythonType valueOf(TypeName typeName){
+        try{
+            return valueOf(typeName.name()); // shouldn't fail
+        }catch (PythonException pe){
+            throw new RuntimeException(pe);
+        }
+
+    }
+    
+    public static final PythonType<String> STR = new PythonType<String>() {
+        @Override
+        public String toJava(PythonObject pythonObject)  throws PythonException{
+            if (!Python.isinstance(pythonObject, Python.strType())){
+                throw new PythonException("Expected variable to be float, but was " +  Python.type(pythonObject));
+            }
+            return pythonObject.toString();
+        }
+        @Override
+        public TypeName getName(){
+            return TypeName.STR;
+        }
+    };
```

Comment:
We can do something like this, 12 -> 6 lines per definition, and less code duplication.
```
    private final TypeName typeName;
    private PythonType(TypeName typeName){
        this.typeName = typeName;
    }

    @Override
    public TypeName getName(){
        return typeName;
    }
    
    protected static void validate(String type, PythonObject pythonObject){
        if (!Python.isinstance(pythonObject, Python.strType())){
            throw new PythonException("Expected variable to be " + type + ", but was " +  Python.type(pythonObject));
        }
    }

    public static final PythonType<String> STR = new PythonType<String>(TypeName.STR) {
        @Override
        public String toJava(PythonObject pythonObject)  throws PythonException{
            validate("String", pythonObject);
            return pythonObject.toString();
        }
    };
```